packages/eslint-plugin/src/rules/prefer-at.ts

```diff
@@ -0,0 +1,219 @@
+        (node.object.type === AST_NODE_TYPES.MemberExpression &&
+          hasCallExpression(node.object))
+      );
+    }
+
+    function getTypeAtLocation(node: TSESTree.Node): ts.Type | undefined {
+      const originalNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (!originalNode) {
+        return undefined;
+      }
+      return checker.getTypeAtLocation(originalNode);
+    }
+
+    type SupportedObject = (type: ts.Type) => boolean;
+
+    function checkObjectName(name: string): SupportedObject {
+      return type => type.getSymbol()?.name === name;
+    }
+
+    function checkObjectType(flags: ts.TypeFlags): SupportedObject {
+      return type => type.getFlags() === flags;
+    }
+
+    const supporterObjects: Array<SupportedObject> = [
+      checkObjectName('Array'),
+      checkObjectName('Int8Array'),
+      checkObjectName('Uint8Array'),
+      checkObjectName('Uint8ClampedArray'),
+      checkObjectName('Int16Array'),
+      checkObjectName('Uint16Array'),
+      checkObjectName('Int32Array'),
+      checkObjectName('Float32Array'),
+      checkObjectName('Uint32Array'),
+      checkObjectName('Float64Array'),
+      checkObjectName('BigInt64Array'),
+      checkObjectName('BigUint64Array'),
+      // eslint-disable-next-line @typescript-eslint/internal/prefer-ast-types-enum
+      checkObjectName('String'),
+      checkObjectType(ts.TypeFlags.String),
+    ];
+
+    function isSupportedObject(type: ts.Type): boolean {
+      return supporterObjects.some(check => check(type));
+    }
+
+    function isExpectedObject(
+      node: TSESTree.Node,
+    ): node is TSESTree.MemberExpression {
+      if (
+        node.type !== AST_NODE_TYPES.MemberExpression ||
+        (options.ignoreFunctions && hasCallExpression(node))
+      ) {
+        return false;
+      }
+      const type = getTypeAtLocation(node.object);
+      if (type && !isSupportedObject(type)) {
+        return false;
+      }
+      const atMember = type?.getProperty('at');
+      return (
+        atMember?.declarations?.every(declaration => {
+          const sourceFile = declaration.getSourceFile();
+          const directory = path.join(sourceFile.fileName, '../');
+          return directory.endsWith(path.join('node_modules/typescript/lib/'));
+        }) ?? false
+      );
+    }
+
+    function isExpectedExpressionLeft(
+      node: TSESTree.BinaryExpression,
+    ): boolean {
+      if (!isExpectedObject(node.left) || getName(node.left) !== 'length') {
+        return false;
+      }
+      const type = getTypeAtLocation(node.left);
+      return type?.getFlags() === ts.TypeFlags.Number;
+    }
+
+    function isExpectedExpressionRight(
+      node: TSESTree.BinaryExpression,
+    ): boolean {
+      const type = getTypeAtLocation(node.right);
+      return (
+        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+        type?.isNumberLiteral() || type?.getFlags() === ts.TypeFlags.Number
+      );
```

Comment:
[Request] We can bypass the lint rule shenanigan with a helpful utility:

```suggestion
      return tsutils.isTypeFlagSet(type, ts.TypeFlags.NumberLike);
```

Anywhere that checks type flags with a `===` should probably use a `tsutils` helper instead. We've seen odd edge cases with number literals vs. numbers before.